File: Leetcode/N-Queens/N_Queens.java
/*
AIM: to place n-queens on an n x n chessboard in such a manner that no two queens attack each other.

ATTACK POSITIONS: same row / same column / same diagonal

ALGORITHM USED : Backtracking
*/

class Solution 
{
    List<List<String>> result = new ArrayList<>(); 	//list of list that provides all possible solutions
    
    public List<List<String>> solveNQueens(int n) 
    {
        int[][] board = new int[n][n];          	//declare and initialize 2D-array 'board'
        for (int i=0; i < n; i++) 
        {
            for (int j=0; j < n; j++) 
            {
                board[i][j] = 0;                	//initialize all the elements with the value 0
            }
        }
        nQueen(board, 0, board.length);         	//call method
        return result;                          	//return required solution
    }

    private boolean nQueen(int[][] board, int row, int N) 
    {
        //Base Case
        if (row == N) 
        {
            constructBoard(board);              	//insert the obtained solution list in the list of solutions
            return false;                       	//Backtrack (to find other possible solutions,if any)
        }
        
        //Consider the given row and check in which column the Queen can be placed
        for (int j=0; j < N; j++) 
        {
            if (isSafe(board, row, j))          	//if current placement of Queen is safe (check by calling the method)
            {
                board[row][j] = 1;              	//place the Queen assuming the current position is safe
                if (nQueen(board, row + 1, N))  	//if the next Queen can also be placed safely
                    return true;
                board[row][j] = 0;              	//remove the Queen since this position is unsafe & Backtrack
            }
        }
        return false;                           	//Could not place Queen in this row
    }

    //method to check if the the current Queen's placement is safe
    private boolean isSafe(int[][] board, int row, int col) 
    {
        //check previous rows of present column
        for(int i=0;i<row;i++) 
        {
            if (board[i][col] == 1)             	//a Queen found
            {
                return false;                   	//unsafe position
            }
        }

        //check for Top Left Diagonal
        int x = row, y = col;
        while (x >= 0 && y >= 0)       
        {
            if (board[x][y] == 1)               	//a Queen found
            {
                return false;                   	//unsafe position
            }
            x--;
            y--;
        }

        //check for Top Right Diagonal
        x = row;
        y = col;
        while (x >= 0 && y < board.length) 
        {
            if (board[x][y] == 1)               	//a Queen found 
            {
                return false;                   	//unsafe position
            }
            x--;
            y++;
        } 
        return true;                            	//current position is safe
    }  

    private void constructBoard(int[][] board) 		//method for adding solution
    {
        List<String> subList = new ArrayList<>();
        for (int[] row : board)                 	//for rows of the board
        {
            StringBuilder sb = new StringBuilder();
            for (int col=0;col<row.length;col++)	//for columns in a row
            {
                if (row[col] == 0) 
                    sb.append(".");             	//empty space
                else 
                    sb.append("Q");             	//Queen placed
            }
            subList.add(sb.toString());         	//one Queen placed in this row
        }
        result.add(subList);                    	//one possible solution added
    }
}

File: Leetcode/N-Queens/N-Queens.py
"""
https://leetcode.com/problems/N-Queens/

The n-queens puzzle is the problem of placing n queens on an n x n chessboard such that no two queens attack each other.

Given an integer n, return all distinct solutions to the n-queens puzzle.
Each solution contains a distinct board configuration of the n-queens' placement, where 'Q' and '.' both indicate a queen and an empty space, respectively.

Example 1:
Input: n = 4
Output:
[[".Q..",
  "...Q",
  "Q...",
  "..Q."],
 ["..Q.",
  "Q...",
  "...Q",
  ".Q.."]]
  
Explanation: There exist two distinct solutions to the 4-queens puzzle as shown above
Example 2:
Input: n = 1
Output: [["Q"]]
"""
from typing import List

class Solution():
    def solveNQueens(self, n) -> List[List[str]]:
        
        # Represent board as matrix of characters to place Queens when needed
        board = [['.' for i in range(n)] for j in range(n)]

        result = []  # Initialize result as empty array

        # Call DFS traversing Column by Column. Starting at column 0.
        self.dfs(board, 0, result)

        # Return result
        return result

    def dfs(self, board, colIndex, result):
        # Terminating condition - to check if we reach the end of the board
        if colIndex == len(board):
            result.append(self.construct(board))
            return

        # In given column, we can put the queen in any row as long as we don't get in the way of other queens on the board
        for i in range(len(board)):
            if self.validate(board, i, colIndex):
                board[i][colIndex] = 'Q'
                self.dfs(board, colIndex + 1, result)
                board[i][colIndex] = '.'

    def validate(self, board, x, y):
        # Goes through the whole board to see if any Queen lies horizontal or diagonal for the Queen you're about to place
        for i in range(len(board)):
            for j in range(y):
                if board[i][j] == 'Q' and (x + j == y + i or x + y == i + j or x == i):
                    return False
        return True


    def construct(self, board):
        # Constructs the result as List[String]
        res = []
        s = ""  # Initailize empty string
        for i in range(len(board)):
            res.append(s.join(board[i]))  # Convert row to string
        return res


if __name__ == '__main__':
    print(Solution().solveNQueens(4))
